From: IrfanMohammad <irfan@graphiant.com>
Date: Fri, 23 Aug 2024 13:59:51 +0000
Subject: Revert "modinfo BUGFIX support union subscription xpaths"

This reverts commit 6730e680fac25a7b270128f4d0dc19b25ef7c74f.

This breaks xpath filtering and oper get callbacks are not called
correctly.
---
 src/common.c  |  22 ++++--------
 src/modinfo.c | 112 ++++++++++++++--------------------------------------------
 2 files changed, 33 insertions(+), 101 deletions(-)

diff --git a/src/common.c b/src/common.c
index 3ab6f75..a87896d 100644
--- a/src/common.c
+++ b/src/common.c
@@ -5340,8 +5340,8 @@ sr_xpath_text_atoms_expr(const char *xpath, const char *prev_atom, const char *e
         uint32_t *atom_count, const char **xpath_next)
 {
     sr_error_info_t *err_info = NULL;
-    uint32_t i, prev_atom_count;
-    int parsed = 0, mod_len, name_len, op_len, last_is_node = 0, atom_stored, len;
+    uint32_t i;
+    int parsed = 0, mod_len, name_len, op_len, last_is_node = 0, len;
     const char *mod, *name, *next, *next2;
     char *tmp, *cur_atom = NULL;
 
@@ -5384,8 +5384,6 @@ sr_xpath_text_atoms_expr(const char *xpath, const char *prev_atom, const char *e
     }
 
 parse_name:
-    atom_stored = 0;
-
     /* parse the name (node, function, ...), there should be some */
     next2 = sr_xpath_next_qname(next, &mod, &mod_len, &name, &name_len);
     if (next2 == next) {
@@ -5427,7 +5425,6 @@ parse_name:
         last_is_node = 0;
     } else if (next[0] == '[') {
         /* predicate(s), get atoms from it */
-        prev_atom_count = *atom_count;
         do {
             next2 = next + 1;
             if ((err_info = sr_xpath_text_atoms_expr(next2, cur_atom, "]", atoms, atom_count, &next))) {
@@ -5446,11 +5443,6 @@ parse_name:
         }
 
         last_is_node = 0;
-
-        if (prev_atom_count < *atom_count) {
-            /* this atom restricted by a predicate has been stored, do not store it without it */
-            atom_stored = 1;
-        }
     }
 
     /* skip whitespaces */
@@ -5461,8 +5453,7 @@ parse_name:
     if ((!end_chars && !next[0]) || (end_chars && next[0] && strchr(end_chars, next[0]))) {
         /* finished with this (sub)expression, add new atom if any found (parsed) */
         parsed = 1;
-        if (!atom_stored && (strlen(prev_atom) < strlen(cur_atom)) &&
-                (err_info = sr_xpath_text_atom_add(&cur_atom, atoms, atom_count))) {
+        if ((strlen(prev_atom) < strlen(cur_atom)) && (err_info = sr_xpath_text_atom_add(&cur_atom, atoms, atom_count))) {
             goto cleanup;
         }
     } else {
@@ -5478,15 +5469,14 @@ parse_name:
         if (op_len) {
             next2 = next + op_len;
             if (last_is_node && end_chars && !strcmp(end_chars, "]") && !strcmp(xpath_ops[i], "=")) {
-                /* check for literal and store it in a special restriction atom (with predicate) */
+                /* check for literal and store it in a special atom */
                 while (isspace(next2[0])) {
                     ++next2;
                 }
                 if ((next2[0] == '\'') || (next2[0] == '\"')) {
                     /* parse and store the literal */
-                    assert(strlen(prev_atom) < strlen(cur_atom));
                     len = (strchr(next2 + 1, next2[0]) - next2) + 1;
-                    if (asprintf(&tmp, "%s[%s=%.*s]", prev_atom, cur_atom + strlen(prev_atom) + 1, len, next2) == -1) {
+                    if (asprintf(&tmp, "%s[.=%.*s]", cur_atom, len, next2) == -1) {
                         SR_ERRINFO_MEM(&err_info);
                         goto cleanup;
                     }
@@ -5509,7 +5499,7 @@ parse_name:
                     }
                     next2 += op_len;
                 }
-            } else if (!atom_stored && (strlen(prev_atom) < strlen(cur_atom))) {
+            } else if (strlen(prev_atom) < strlen(cur_atom)) {
                 /* add a new atom if a new one */
                 if ((err_info = sr_xpath_text_atom_add(&cur_atom, atoms, atom_count))) {
                     goto cleanup;
diff --git a/src/modinfo.c b/src/modinfo.c
index f94f1b2..dae64fc 100644
--- a/src/modinfo.c
+++ b/src/modinfo.c
@@ -812,62 +812,7 @@ sr_modinfo_changesub_rdunlock(struct sr_mod_info_s *mod_info)
 }
 
 /**
- * @brief Check whether operational data request and subscription predicate match.
- *
- * @param[in] request_pred Request predicate.
- * @param[in] sub_pred Subscription predicate.
- * @return 0 predicate for different nodes, not applicable;
- * @return 1 predicate for the same nodes with the same value, matches;
- * @return 2 predicate for the same nodes with different values, does not match.
- */
-static int
-sr_xpath_oper_data_text_atom_pred_match(const char *request_pred, const char *sub_pred)
-{
-    const char *req_ptr, *sub_ptr, *mod1, *name1, *val1, *mod2, *name2, *val2;
-    int mlen1, mlen2, len1, len2;
-
-    req_ptr = request_pred;
-    sub_ptr = sub_pred;
-
-    do {
-        /* parse nodes */
-        req_ptr = sr_xpath_next_qname(req_ptr + 1, &mod1, &mlen1, &name1, &len1);
-        sub_ptr = sr_xpath_next_qname(sub_ptr + 1, &mod2, &mlen2, &name2, &len2);
-
-        /* module name */
-        if ((mlen1 && mlen2) && ((mlen1 != mlen2) || strncmp(mod1, mod2, mlen1))) {
-            /* different modules */
-            return 0;
-        }
-
-        /* node name */
-        if ((len1 != len2) || strncmp(name1, name2, len1)) {
-            /* different node names */
-            return 0;
-        }
-    } while ((req_ptr[0] != '=') && (sub_ptr[0] != '='));
-
-    if ((req_ptr[0] != '=') || (sub_ptr[0] != '=')) {
-        /* path continues */
-        return 0;
-    }
-
-    ++req_ptr;
-    ++sub_ptr;
-
-    /* compare values */
-    val1 = req_ptr + 1;
-    len1 = strchr(val1, req_ptr[0]) - val1;
-    val2 = sub_ptr + 1;
-    len2 = strchr(val2, sub_ptr[0]) - val2;
-    if (len1 != len2) {
-        return 2;
-    }
-    return strncmp(val1, val2, len1) ? 2 : 1;
-}
-
-/**
- * @brief Check whether operational data are required based on a single request and subscription atom.
+ * @brief Check whether operational data are required based on single request and subscription atom.
  *
  * @param[in] request_atom Request text atom.
  * @param[in] sub_atom Subscription text atom.
@@ -878,7 +823,7 @@ sr_xpath_oper_data_text_atom_pred_match(const char *request_pred, const char *su
 static int
 sr_xpath_oper_data_text_atoms_required(const char *request_atom, const char *sub_atom)
 {
-    const char *req_ptr, *sub_ptr, *mod1, *name1, *mod2, *name2;
+    const char *req_ptr, *sub_ptr, *mod1, *name1, *val1, *mod2, *name2, *val2;
     int mlen1, mlen2, len1, len2, wildc1, wildc2;
 
     req_ptr = request_atom;
@@ -913,18 +858,16 @@ sr_xpath_oper_data_text_atoms_required(const char *request_atom, const char *sub
             return 0;
         }
 
-        /* predicate, always at the end and so decides the match */
+        /* value */
         if ((req_ptr[0] == '[') && (sub_ptr[0] == '[')) {
-            if (sr_xpath_oper_data_text_atom_pred_match(req_ptr + 1, sub_ptr + 1) == 2) {
-                /* filtered out */
+            val1 = req_ptr + 4;
+            len1 = strchr(val1, req_ptr[3]) - val1;
+            val2 = sub_ptr + 4;
+            len2 = strchr(val2, sub_ptr[3]) - val2;
+            if ((len1 != len2) || strncmp(val1, val2, len1)) {
+                /* different values, filtered out */
                 return 2;
-            } else {
-                /* different nodes or the value matches */
-                return 1;
             }
-        } else if ((req_ptr[0] == '[') || (sub_ptr[0] == '[')) {
-            /* predicate only in one atom and it must end after the predicate */
-            break;
         }
 
         /* parse until the subscription path ends */
@@ -963,7 +906,7 @@ sr_xpath_oper_data_required(const char *request_xpath, const char *sub_xpath, in
 {
     sr_error_info_t *err_info = NULL;
     sr_xp_atoms_t *req_atoms = NULL, *sub_atoms = NULL;
-    uint32_t i, j, k, l;
+    uint32_t i, j, k;
     int r;
 
     assert(sub_xpath);
@@ -982,33 +925,32 @@ sr_xpath_oper_data_required(const char *request_xpath, const char *sub_xpath, in
     if ((err_info = sr_xpath_get_text_atoms(sub_xpath, &sub_atoms)) || !sub_atoms) {
         goto cleanup;
     }
+    assert(sub_atoms->union_count == 1);
 
     /* check whether any atoms match */
     *required = 0;
     for (i = 0; i < req_atoms->union_count; ++i) {
-        for (j = 0; j < sub_atoms->union_count; ++j) {
-            for (k = 0; k < req_atoms->unions[i].atom_count; ++k) {
-                for (l = 0; l < sub_atoms->unions[j].atom_count; ++l) {
-                    r = sr_xpath_oper_data_text_atoms_required(req_atoms->unions[i].atoms[k], sub_atoms->unions[j].atoms[l]);
-                    if (r == 1) {
-                        /* required but need to check all atoms */
-                        *required = 1;
-                    } else if ((r == 0) || (r == 2)) {
-                        /* not required for the union */
-                        *required = 0;
-                        break;
-                    }
-                }
-                if (l < sub_atoms->unions[j].atom_count) {
+        for (j = 0; j < req_atoms->unions[i].atom_count; ++j) {
+            for (k = 0; k < sub_atoms->unions[0].atom_count; ++k) {
+                r = sr_xpath_oper_data_text_atoms_required(req_atoms->unions[i].atoms[j], sub_atoms->unions[0].atoms[k]);
+                if (r == 1) {
+                    /* required but need to check all atoms */
+                    *required = 1;
+                } else if (r == 2) {
+                    /* not required for the union */
+                    *required = 0;
                     break;
                 }
             }
-
-            if (*required) {
-                /* required for a union */
-                goto cleanup;
+            if (k < sub_atoms->unions[0].atom_count) {
+                break;
             }
         }
+
+        if (*required) {
+            /* required for a union */
+            goto cleanup;
+        }
     }
 
 cleanup:
